src/server/pfs/server/commit_set.go

```diff
@@ -0,0 +1,497 @@
+			return nil
+		}
+		sent[pfsdb.CommitKey(ci.Commit)] = struct{}{}
+		return cb(ci)
+
+	}
+	unfinishedCommits := make([]*pfs.Commit, 0)
+	// NOTE: before 2.5, a triggered commits would be included as part of the same commit set as the triggering commit set,
+	// so we would wait for all of the current commit set to finish, then check if new previously unknown commits are added due to triggers.
+	if err := d.inspectCommitSetImmediate(ctx, commitset, func(ci *pfs.CommitInfo) error {
+		if ci.Finished != nil {
+			return send(ci)
+		}
+		unfinishedCommits = append(unfinishedCommits, proto.Clone(ci.Commit).(*pfs.Commit))
+		return nil
+	}); err != nil {
+		return err
+	}
+	for _, uc := range unfinishedCommits {
+		// TODO: make a dedicated call just for the blocking part, inspectCommit is a little heavyweight?
+		ci, err := d.inspectCommit(ctx, uc, pfs.CommitState_FINISHED)
+		if err != nil {
+			return err
+		}
+		if err := send(ci); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *driver) listCommitSet(ctx context.Context, project *pfs.Project, cb func(*pfs.CommitSetInfo) error) error {
+	// Track the commitsets we've already processed
+	seen := map[string]struct{}{}
+	// Return commitsets by the newest commit in each set (which can be at a different
+	// timestamp due to triggers or deferred processing)
+	commitInfo := &pfs.CommitInfo{}
+	err := d.commits.ReadOnly(ctx).List(commitInfo, col.DefaultOptions(), func(string) error {
+		if commitInfo.GetCommit().GetBranch().GetRepo().GetProject().GetName() != project.GetName() {
+			return nil
+		}
+		if _, ok := seen[commitInfo.Commit.ID]; ok {
+			return nil
+		}
+		seen[commitInfo.Commit.ID] = struct{}{}
+		var commitInfos []*pfs.CommitInfo
+		err := d.inspectCommitSet(ctx, &pfs.CommitSet{ID: commitInfo.Commit.ID}, false, func(ci *pfs.CommitInfo) error {
+			commitInfos = append(commitInfos, ci)
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+		return cb(&pfs.CommitSetInfo{
+			CommitSet: client.NewCommitSet(commitInfo.Commit.ID),
+			Commits:   commitInfos,
+		})
+	})
+	return errors.EnsureStack(err)
+}
+
+// dropCommitSet is only implemented for commits with no children, so if any
+// commits in the commitSet have children the operation will fail.
+func (d *driver) dropCommitSet(txnCtx *txncontext.TransactionContext, commitset *pfs.CommitSet) error {
+	css, err := d.subvenantCommitSets(txnCtx, commitset)
+	if err != nil {
+		return err
+	}
+	if len(css) > 0 {
+		return &pfsserver.ErrSquashWithSubvenance{CommitSet: commitset, SubvenantCommitSets: css}
+	}
+	cis, err := d.inspectCommitSetImmediateTx(txnCtx, commitset, true)
+	if err != nil {
+		return err
+	}
+	for _, ci := range cis {
+		if ci.Commit.Branch.Repo.Type == pfs.SpecRepoType && ci.Origin.Kind == pfs.OriginKind_USER {
+			return errors.Errorf("cannot squash commit %s because it updated a pipeline", ci.Commit)
+		}
+		// TODO(acohen4): can drop commits & squash just be modeled the same for now?
```

Comment:
No, a drop is interpreted as removing the data from the subsequent commits, so keeping it limited to the no child case is the way to go.